src/renderer/Renderer.js

```diff
@@ -216,8 +216,9 @@ export default class Renderer {
         gl.depthMask(false);
         gl.bindFramebuffer(gl.FRAMEBUFFER, this.auxFB);
 
-        const styleDataframe = (dataframe, dataframeTexture, shader, vizExpr) => {
-            const textureId = shader.textureIds.get(viz);
+        const styleDataframe = (dataframe, dataframeTexture, metashader, vizExpr) => {
```

Comment:
This function only uses `this.gl`, so it can be extracted with a refactor to something like this:

```js
_styleDataframe (dataframe, dataframeTexture, metashader, vizExpr) => {
            const gl = this.gl;

            const shader = metashader.shader;
            const textureId = metashader.textureIds;

            gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, dataframeTexture, 0);
            gl.viewport(0, 0, RTT_WIDTH, dataframe.height);
            gl.clear(gl.COLOR_BUFFER_BIT);

            gl.useProgram(shader.program);
            // Enforce that property texture TextureUnit don't clash with auxiliar ones
            drawMetadata.freeTexUnit = Object.keys(textureId).length;
            vizExpr._setTimestamp((Date.now() - INITIAL_TIMESTAMP) / 1000.0);
            vizExpr._preDraw(shader.program, drawMetadata, gl);

            Object.keys(textureId).forEach((name, i) => {
                gl.activeTexture(gl.TEXTURE0 + i);
                gl.bindTexture(gl.TEXTURE_2D, dataframe.getPropertyTexture(name));
                gl.uniform1i(textureId[name], i);
            });

            gl.enableVertexAttribArray(shader.vertexAttribute);
            gl.bindBuffer(gl.ARRAY_BUFFER, this.bigTriangleVBO);
            gl.vertexAttribPointer(shader.vertexAttribute, 2, gl.FLOAT, false, 0, 0);

            gl.drawArrays(gl.TRIANGLES, 0, 3);
            gl.disableVertexAttribArray(shader.vertexAttribute);
        };
```

Reply:
:+1:

Reply:
I was already doing the refactor when I remembered why this was left there, increasing considerably the already long render method.

The reason is that the code relies a lot in the current state of WebGL, therefore the function is completely tied to that place. It depends on a lot of state set to WebGL before its call, and it probably lefts the WebGL state in a particular state.

Therefore, even though extracting it is trivial, there are a lot of dependencies (WebGL side effects) that are needed.